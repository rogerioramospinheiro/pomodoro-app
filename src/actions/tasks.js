export const startTask = ({title, description, start_date, activity_time, rest_time} = {}) => ({
    type: 'START_TASK',
    task: {
        title,
        description,
        start_date,
        activity_time,
        rest_time
    }
});

export const cancelTask = () => ({
    type: 'CANCEL_TASK'
});

export const finishTask = ({end_date} = {}) => ({
    type: 'FINISH_TASK',
    task: {
        end_date
    }
});