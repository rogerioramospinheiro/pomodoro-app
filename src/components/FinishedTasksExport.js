import React from 'react';
import { CSVLink } from 'react-csv';
import moment from 'moment';

const FinishedTasksExport = ({tasks}) => {
    const headers = ['ID', 'TITLE', 'DESCRIPTION', 'START TIME', 'END TIME', 'DURATION'];
    const csvData = [headers];
    tasks.forEach( (task) => {
        const start_moment = moment(task.start_date);
        const end_moment = moment(task.end_date);
        const row = [ 
            task.id, 
            task.title, 
            task.description, 
            start_moment.format('hh:mm:ss A'), 
            end_moment.format('hh:mm:ss A'),
            `${end_moment.diff(start_moment, 'minutes')}m`
        ]
        csvData.push(row);
    });
    const printDate = moment(tasks[0].start_date).format('YYYYMMDD');
    const fileName = `pomodoro-app-export-${printDate}.csv`
    return (
        <CSVLink data={csvData} filename={fileName} enclosingCharacter={`"`} className="finished-tasks__link">
            <div className="task-form__button finished-tasks__link-export button-wide">Export</div>
        </CSVLink>
    );
};

export default FinishedTasksExport;