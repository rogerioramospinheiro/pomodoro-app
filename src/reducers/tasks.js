import uuidv1 from 'uuid/v1';

const tasksReducerDefaultState = {
    state_name: 'IDLE',
    current_task: {
    },
    finished_tasks: []
};

const transitionStartToCounting = (state, action) => {
    if (state.state_name === 'IDLE') {
        return {
            state_name: 'COUNTING',
            current_task: {
                title: action.task.title,
                description: action.task.description,
                start_date: action.task.start_date,
            },
            finished_tasks: [...state.finished_tasks]
        };
    } else {
        return state;
    }
};

const transitionCancelToIdle = (state) => {
    if (state.state_name !== 'IDLE') {
        return {
            state_name: 'IDLE',
            current_task: tasksReducerDefaultState.current_task,
            finished_tasks: [...state.finished_tasks]
        }
    } else {
        return state;
    }
};

const transitionFinishToIdle = (state, action) => {
    if (state.state_name === 'COUNTING') {
        return {
            state_name: 'RESTING',
            current_task: {
                title: state.current_task.title,
                description: state.current_task.description,
                start_date: state.current_task.start_date,
            },
            finished_tasks: [
                ...state.finished_tasks
            ]
        }
    } else if (state.state_name === 'RESTING') {
        const finished_task = {
            id: uuidv1(),
            title: state.current_task.title,
            description: state.current_task.description,
            start_date: state.current_task.start_date,
            end_date: action.task.end_date
        };
        return {
            state_name: 'IDLE',
            current_task: tasksReducerDefaultState.current_task,
            finished_tasks: [
                finished_task,
                ...state.finished_tasks
            ]
        }
    } else {
        return state;
    }
}

const completeEditFinishedTask = (state, action) => {
    const finished_tasks = state.finished_tasks.map( (task) => {
        if (task.id === action.id) {
            return {
                ...task,
                ...action.updates
            };
        } else {
            return task;
        }
    });
    return {
        ...state,
        finished_tasks
    };
}

export default (state = tasksReducerDefaultState, action) => {
    switch(action.type) {
        case 'START_TASK':
            return transitionStartToCounting(state, action);
        case 'CANCEL_TASK':
            return transitionCancelToIdle(state);
        case 'FINISH_TASK':
            return transitionFinishToIdle(state, action);
        case 'EDIT_FINISHED_TASK':
            return completeEditFinishedTask(state, action);
        default:
            return state;
    }
};