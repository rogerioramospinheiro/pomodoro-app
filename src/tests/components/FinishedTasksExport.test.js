import React from 'react';
import { shallow } from 'enzyme';
import FinishedTasksExport from '../../components/FinishedTasksExport';
import { createFinishedTasksState } from '../fixtures/tasks';

test('Should render FinishedTasksExport', () => {
    const state = createFinishedTasksState();
    const tasks = state.finished_tasks;
    const wrapper = shallow(<FinishedTasksExport tasks={tasks} />);
    expect(wrapper).toMatchSnapshot();
});