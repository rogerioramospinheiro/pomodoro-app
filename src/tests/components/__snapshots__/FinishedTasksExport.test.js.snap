// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render FinishedTasksExport 1`] = `
<CSVLink
  asyncOnClick={false}
  className="finished-tasks__link"
  data={
    Array [
      Array [
        "ID",
        "TITLE",
        "DESCRIPTION",
        "START TIME",
        "END TIME",
        "DURATION",
      ],
      Array [
        "c638040c-3917-11ea-a137-2e728ce88125",
        "TASK-0001",
        "Job to be finished soon",
        "07:00:10 PM",
        "07:00:30 PM",
        "0m",
      ],
      Array [
        "cf20da8a-3917-11ea-a137-2e728ce88125",
        "TASK-0002",
        "Job to finish by the end of the day",
        "07:00:40 PM",
        "07:01:00 PM",
        "0m",
      ],
      Array [
        "d79e2000-3917-11ea-a137-2e728ce88125",
        "TASK-0003",
        "Job to be finished tomorrow",
        "07:01:10 PM",
        "07:01:40 PM",
        "0m",
      ],
    ]
  }
  enclosingCharacter="\\""
  filename="pomodoro-app-export-19691231.csv"
  separator=","
  target="_blank"
  uFEFF={true}
>
  <div
    className="task-form__button finished-tasks__link-export button-wide"
  >
    Export
  </div>
</CSVLink>
`;
